Unable to perform copydb (panic: runtime error: memory)

#### System information

Geth version: `Geth/v1.7.3-stable-4bb3c89d/linux-amd64/go1.9`
OS & Version: Ubuntu 16.04.3 x64

I was following @karalabe 's instructions in #15797 and I got to block number 2433095 when geth crashed:

Command I ran `geth --datadir=/gethswap copydb --cache=512 /geth/.ethereum/geth/chaindata`



````
INFO [01-11|15:34:00] Imported new block receipts              count=283  elapsed=128.139ms bytes=817740   number=2433095 hash=65693e…3957ec ignored=0

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa45f88]

goroutine 58 [running]:
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).fetchHeaders(0xc420216680, 0xc42025ce10, 0x258001, 0x0, 0x0)
        /build/ethereum-ZSPmrv/ethereum-1.7.3+build11486+xenial/build/_workspace/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:859 +0xf18
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).syncWithPeer.func3(0x8, 0xffd8b0)
        /build/ethereum-ZSPmrv/ethereum-1.7.3+build11486+xenial/build/_workspace/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:474 +0x40
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync.func1(0xc42c1a5ee0, 0xc420062c60, 0xc42851f3a0)
        /build/ethereum-ZSPmrv/ethereum-1.7.3+build11486+xenial/build/_workspace/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:500 +0x51
created by github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync
        /build/ethereum-ZSPmrv/ethereum-1.7.3+build11486+xenial/build/_workspace/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:500 +0xc7

root@GETH01:/# free -h
              total        used        free      shared  buff/cache   available
Mem:           7.8G         64M        2.3G        516K        5.4G        7.4G
Swap:          4.0G         30M        4.0G

root@GETH01:/# df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            3.9G     0  3.9G   0% /dev
tmpfs           797M  8.8M  788M   2% /run
/dev/sda1       193G  174G  9.8G  95% /
tmpfs           3.9G     0  3.9G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           3.9G     0  3.9G   0% /sys/fs/cgroup
/dev/sdb1        99G  4.6G   89G   5% /gethswap
tmpfs           797M     0  797M   0% /run/user/1000
````

I stumbled upon #15797 too and tried to get geth to "fast sync" with itself. It seems to have caused the same problem here.
This feature seemed just too good to be true! Or at least too good for people with small SSDs which is my case.

In my case it crashed around block 4253696. I happened to run the command with more cache (1024) than the OP. I will run the command with 2048 cache and report back the results after it's done.  Full command: ```"C:\Ethereum Wallet\binaries\Geth\unpacked\geth.exe" --datadir=M:\temp_ethereum_swap copydb --cache=1024 C:\Ethereum\geth\chaindata```

geth version output:
```
Version: 1.7.2-stable
Git Commit: 1db4ecdc0b9e828ff65777fb466fc7c1d04e0de9
Architecture: amd64
Protocol Versions: [63 62]
Network Id: 1
Go Version: go1.9
Operating System: windows
GOPATH=
GOROOT=c:\go
```
OS: Windows 10 Pro

End of the output:
```
INFO [01-29|16:15:03] Imported new block headers               count=2048 elapsed=3.481s     number=4253696 hash=e3597f…fb2161 ignored=0
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0xa41cb7]

goroutine 66 [running]:
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).fetchParts(0xc0420529c0, 0x15c9820, 0xc04202b340, 0xc0420386c0, 0xc1afc740a0, 0xc1a82c7490, 0xc0938439e8, 0xc093843a28, 0xc093843a18, 0xfbcc78, ...)
        C:/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:1074 +0x657
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).fillHeaderSkeleton(0xc0420529c0, 0x41a001, 0xc0cecde400, 0x80, 0x80, 0x4, 0x6000, 0x6000, 0x0, 0x0, ...)
        C:/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:904 +0x384
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).fetchHeaders(0xc0420529c0, 0xc0421810e0, 0x41a001, 0x0, 0x0)
        C:/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:835 +0xa74
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).syncWithPeer.func3(0x8, 0xfbee20)
        C:/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:474 +0x47
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync.func1(0xc05a520450, 0xc05d2386c0, 0xc04223dfc0)
        C:/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:500 +0x58
created by github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync
        C:/gopath/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:500 +0xce
```

--- a/eth/downloader/downloader.go
+++ b/eth/downloader/downloader.go
@@ -324,8 +324,13 @@ func (d *Downloader) Synchronise(id string, head common.Hash, td *big.Int, mode 
 		errEmptyHeaderSet, errPeersUnavailable, errTooOld,
 		errInvalidAncestor, errInvalidChain:
 		log.Warn("Synchronisation failed, dropping peer", "peer", id, "err", err)
-		d.dropPeer(id)
-
+		if d.dropPeer == nil {
+			// The dropPeer method is nil when `--copydb` is used for a local copy.
+			// Timeouts can occur if e.g. compaction hits at the wrong time, and can be ignored
+			log.Warn("Downloader wants to drop peer, but peerdrop-function is not set", "peer", id)
+		} else {
+			d.dropPeer(id)
+		}
 	default:
 		log.Warn("Synchronisation failed, retrying", "err", err)
 	}
@@ -853,6 +858,12 @@ func (d *Downloader) fetchHeaders(p *peerConnection, from uint64) error {
 			getHeaders(from)
 
 		case <-timeout.C:
+			if d.dropPeer == nil {
+				// The dropPeer method is nil when `--copydb` is used for a local copy.
+				// Timeouts can occur if e.g. compaction hits at the wrong time, and can be ignored
+				p.log.Warn("Downloader wants to drop peer, but peerdrop-function is not set", "peer", p.id)
+				break
+			}
 			// Header retrieval timed out, consider the peer bad and drop
 			p.log.Debug("Header request timed out", "elapsed", ttl)
 			headerTimeoutMeter.Mark(1)
@@ -1071,7 +1082,13 @@ func (d *Downloader) fetchParts(errCancel error, deliveryCh chan dataPack, deliv
 						setIdle(peer, 0)
 					} else {
 						peer.log.Debug("Stalling delivery, dropping", "type", kind)
-						d.dropPeer(pid)
+						if d.dropPeer == nil {
+							// The dropPeer method is nil when `--copydb` is used for a local copy.
+							// Timeouts can occur if e.g. compaction hits at the wrong time, and can be ignored
+							peer.log.Warn("Downloader wants to drop peer, but peerdrop-function is not set", "peer", pid)
+						} else {
+							d.dropPeer(pid)
+						}
 					}
 				}
 			}

